### ISSUE TITLE
Magic Numbers cop proposal

### ISSUE THREAD

**Issue by maasha** [bbatsov/rubocop#1975]
Is it possible to write a cop to detect [Magic Numbers](https://en.wikipedia.org/wiki/Magic_number_(programming))?:

> Unique values with unexplained meaning or multiple occurrences which could (preferably) be replaced with named constants

It might be overkill for incrementations of 1:

    count += 1

but other places it will be in order:

    DAYS_IN_WEEK = 7

    days += DAYS_IN_WEEK

**Comment by bankair** [comment]
You probably want to exclude 0 as well, for example:

```
puts 'balance positive' if my_sum > 0
```

**Comment by maasha** [comment]
In this case 8 is replaced with a constant, but not so for 1, which makes sense IMHO.

    BITS_IN_BYTE = 8

    mask = (1 << BITS_IN_BYTE) - 1

**Comment by bbatsov** [comment]
If you take a look at the specs of RuboCop you'll see plenty of uses of numbers to check the size of anything. There's little point in having them as constants.

**Comment by maasha** [comment]
So where does it make sense?

**Comment by bbatsov** [comment]
I'm just saying it's not easy to come up with a check that will yield no false positives.

**Comment by maasha** [comment]
Fair enough. Lets start with the easy ones: Which are clearly magic numbers and which are not.

**Comment by mikegee** [comment]
I want to support this proposal. I'm keen to extract magic numbers in my code and suggest it in reviews.

But how can @bbatsov's concern be addressed? Literal numeric values frequently appear in tests and should not be extracted to a constant. Should Rubocop ignore tests and specs somehow?

**Comment by fj** [comment]
How about this proposal for the rule?

Any numeric literal

* appearing in a comparison expression (`>`, `==`, `eql`, `<`, etc.)
* whose values are other than -1, 0, or 1

is flagged as a magic number?

This is suitably conservative, I think. For example, it won't trigger on @maasha's example of `mask = (1 << BITS_IN_BYTE) - 1`. It also won't trigger in RSpec tests because the literals are usually supplied to an assert expression, and not used in a comparison.

**Comment by mikegee** [comment]
@fj That makes enough sense. I wonder who has enough know-how or energy to put together a PR? Getting it working on Rubocop itself would be a significant achievement.

**Comment by bbatsov** [comment]
If anyone wants to work on this - please, step forward. Otherwise I'll close this ticket for lack of activity.

**Comment by Bodacious** [comment]
For anyone still interested in this feature, my colleague and I have published this gem:
https://github.com/bodacious/rubocop-magic_numbers/

Feedback welcome

**Comment by Bodacious** [comment]
I would be happy to, if there was interest from the Rubocop team?  @bbatsov wdyt?
